File: 001-020/014.js
// 014_another.jsのプログラムの方が早いが、
// 今回のような問題だと、配列の引数の上限を超えるため、
// 正しく計算できないことがある。その時はこちらの方がいい
function collatz_length(num) {
  let length = 1;
  let calc = num;
  while (calc !== 1) {
    length++;
    if (calc % 2 === 0) {
      calc /= 2;
    } else {
      calc = calc * 3 + 1;
    }
  }
  return length;
}

//初項が1,000,000未満の数の時、どの数から始めれば最長の数列を生成できるか

let maximum = 0;
let num = 0;
for (let i = 999999; i > 1; i-- ){
  let tmp = collatz_length(i);
  if (tmp > maximum){
    maximum = tmp;
    num = i;
  }
}

console.log(num); // 解答
console.log(maximum);

File: 001-020/014.rb
# ruby で一つ一つ計算しているとかなり時間がかかるので、
# メモリを活用する
# メモリを一切使わないときと比較して10倍の速さで計算できる
maximum = 0
length = 0
coll = [0, 1, 2]

3.upto(1_000_000) do |i|
  tmp = 0
  calc = i
  while calc > coll.length - 1
    tmp += 1
    calc = calc.even? ? calc / 2 : calc * 3 + 1
  end
  tmp += coll[calc]
  coll << tmp
  if tmp > length
    length = tmp
    maximum = i
  end
end

puts maximum # 解答
puts length
